librz/bin/dbginfo.c

```diff
@@ -239,13 +239,104 @@ RZ_DEPRECATE RZ_API bool rz_bin_addr2line(RzBin *bin, ut64 addr, char *file, int
 	return false;
 }
 
-RZ_DEPRECATE RZ_API char *rz_bin_addr2text(RzBin *bin, ut64 addr, int origin) {
+static const char *cache_lines(RzBinSourceLineCacheItem *x, ut64 line) {
+	rz_warn_if_fail(line > x->cached_lines);
+
+	if (!x->file_content) {
+		return NULL;
+	}
+	bool found = false;
+	char *str = ht_up_find(x->line_by_ln, x->cached_lines, &found);
+	if (!(found && str)) {
+		rz_warn_if_reached();
+		return NULL;
+	}
+
+	ut64 i = x->cached_lines;
+	char *p = str;
+	char *q = NULL;
+	do {
+		q = strchr(p, '\n');
+		++i;
+		if (q) {
+			p = q + 1;
+			*q = '\0';
+			ht_up_update(x->line_by_ln, i, q + 1);
+			x->cached_lines = i;
+		}
+	} while ((p && p - x->file_content < x->file_size));
+
+	str = ht_up_find(x->line_by_ln, line, &found);
+	if (!(found && str)) {
+		return NULL;
+	}
+	q = rz_str_trim_nc(str);
+	if (q != str) {
+		ht_up_update(x->line_by_ln, line, q);
+	}
+	return q;
+}
+
+static const char *read_line(const char *file, int line, RzBinSourceLineCache *cache) {
+	rz_return_val_if_fail(file, NULL);
+	if (!(cache && cache->items)) {
+		rz_file_slurp_line(file, line, 0);
+	}
+	bool found = false;
+	char *content = NULL;
+	size_t sz = 0;
+	RzBinSourceLineCacheItem *item = ht_pp_find(cache->items, file, &found);
+	if (found) {
+		if (!(item && item->file_content)) {
+			return NULL;
+		} else {
+			char *str = ht_up_find(item->line_by_ln, line, NULL);
+			if (!str) {
+				return NULL;
+			}
+			char *q = rz_str_trim_nc(str);
+			if (q != str) {
+				ht_up_update(item->line_by_ln, line, q);
+			}
+			return q;
+		}
+	} else {
+		content = rz_file_slurp(file, &sz);
+		if (!content) {
+			ht_pp_insert(cache->items, file, NULL);
+			return NULL;
+		}
+		item = RZ_NEW0(RzBinSourceLineCacheItem);
+		if (!item) {
+			goto err;
+		}
+		item->file_content = content;
+		item->file_size = sz;
+		item->line_by_ln = ht_up_new0();
+		if (!item->line_by_ln) {
+			goto err;
+		}
+		item->cached_lines = 1;
+		ht_up_update(item->line_by_ln, 1, item->file_content);
+		ht_pp_update(cache->items, file, item);
+		return cache_lines(item, line);
+	}
+err:
+	if (item) {
+		ht_up_free(item->line_by_ln);
+	}
+	free(content);
+	free(item);
+	return NULL;
+}
+
+RZ_DEPRECATE RZ_API char *rz_bin_addr2text(RzBin *bin, ut64 addr, RzDebugInfoOption opt) {
```

Comment:
I suppose the deprecation comes from the fact that it uses `RzBin` and we get rid of the functions accepting only `RzBin` - they should use `RzBinFile` or `RzBinObject` depending on the application since one `RzBin` could contain multiple instances of those, just see how other `rz_bin_get_*()` APIs were deprecated: https://github.com/rizinorg/rizin/issues/3755

Reply:
Please also add ownership of the return value